### ISSUE TITLE
[Style] Support backslashy method modifiers (the python decorator way)

### ISSUE THREAD

**Issue by vmeyet** [bbatsov/rubocop#1937]
I'd like to use the following syntax (possible since ruby 2.2), which is not compatible with rubocop
```ruby
# memoize is a method modifier
memoize \
def testme
  "Hello"
end

# instead of
def testme
  "Hello"
end
memoize :testme
```

Upside:
- `def` stays aligned with all the other `def`
- DRY (method name is already returned by `def` in ruby 2.2)
- feels like python decorators

Is there any way to use the existing rules, to allow it (without disabling rules)?
Will you support that syntax at some point?

(see http://thepugautomatic.com/2015/01/backslashy-ruby/)

**Comment by bbatsov** [comment]
I think that's a bad idea, as it's high unidiomatic (and abuses `/`). A common approach for this is:

```ruby
memoize def testme
  "Hello"
end
```

**Comment by vmeyet** [comment]
I know but it's not applicable multiple time (or really ugly), and it hides the `def` part (personal taste).

ex
```ruby
def testme
  "Hello"
end
memoize :testme
wrapme :testme
somehelper :testme

# vs

somehelper wrapme memoize def testme
  "Hello"
end

# vs

somehelper \
wrapme \
memoize \
def testme
  "Hello"
end
```

But anyway, even with the syntax you're proposing, rubocop is making warnings. Will this (the syntax you proposed) be supported ?

**Comment by rrosenblum** [comment]
I decided to play around with this last night since I had never seen this syntax before. Here are the contents of my test file

```ruby
# encoding: utf-8
memoize \
  def testme
    'Hello'
  end

# instead of
def testme
  'Hello'
end
memoize :testme

# this is documentation
class Test
  memoize \
    def slash
      'Hello'
    end

  # instead of
  def testme
    'Hello'
  end
  memoize :testme

  private_class_method \
    def self.foo
    end

  helper_method \
    def current_user
    end
end
```

@vmeyet are you asking if RuboCop can suggest to use this style, or are you asking for RuboCop to not produce offenses/errors when processing this style?

In my opinion, since this is valid syntax, RuboCop should be able to process this code without error. That being said, RuboCop produces the following output when I run it against the example file.

```
$ b rubocop -D test.rb
Inspecting 1 file
An error occurred while Style/IndentationWidth cop was inspecting /Users/rrosenblum/work/rubocop/test.rb.
To see the complete backtrace run rubocop -d.
W

Offenses:

test.rb:28:5: W: Lint/DefEndAlignment: end at 28, 4 is not aligned with `private_class_method \
    def` at 27, 4
    end
    ^^^

1 file inspected, 1 offense detected

1 error occurred:
An error occurred while Style/IndentationWidth cop was inspecting /Users/rrosenblum/work/rubocop/test.rb.
Errors are usually caused by RuboCop bugs.
Please, report your problems to RuboCop's issue tracker.
Mention the following information in the issue report:
0.31.0 (using Parser 2.2.2.5, running on ruby 2.2.2 x86_64-darwin14)
```

I will defer to @bbatsov and @jonas054 for any decisions being made, about the style and whether or not to have a cop for it. It does seem though, that we may want to look into `Style/IndentationWidth` and `Lint/DefEndAlignment`

**Comment by vmeyet** [comment]
> @vmeyet are you asking if RuboCop can suggest to use this style, or are you asking for RuboCop to not produce offenses/errors when processing this style?

I'm asking to not produce offenses/errors or (better) having a rule where you can set an `EnforcedStyle` options.
(this could be an offense by default, but with the `EnforcedStyle: backslash_method_modifier` option, it will enforce that style).

**Comment by bbatsov** [comment]
Closing due to lack of activity.
